Framework/src/Framework/HistogramPool.cxx

```diff
@@ -0,0 +1,299 @@
+            "\nMake sure to `histograms_.create` in onProcessStart for any "
+            "histogram you want to `histograms_.fill`.");
+  }
+
+  return histograms_[name];
+}
+
+void HistogramPool::insert(const std::string& name,
+                           std::function<TH1*()> factory, bool weighted) {
+  if (histograms_.find(name) != histograms_.end()) {
+    EXCEPTION_RAISE(
+        "RepeatName",
+        "Histogram " + name +
+            " already exists in histogram pool."
+            "\nMake sure to use distinct names for your different histograms!"
+            "\nYou can use `histograms_.get` to retrieve a pointer to a "
+            "specific histogram "
+            "if you want to do some other customizations besides filling.");
+  }
+
+  get_directory_()->cd();
+
+  auto h = factory();
+  if (weighted) h->Sumw2();
+  histograms_[name] = h;
+}
+
+std::tuple<std::size_t, double, double> categoryBins(
+    const std::vector<std::string>& categories, int offset = 0) {
+  std::size_t n_categories = categories.size();
+  double min = offset - 0.5;
+  double max = offset + n_categories + 0.5;
+  return std::make_tuple(n_categories, min, max);
+}
+
+void labelAxis(TAxis* axis, const std::vector<std::string>& categories) {
+  for (std::size_t ibin{1}; ibin <= categories.size(); ibin++) {
+    axis->SetBinLabel(ibin, categories[ibin - 1].c_str());
+  }
+}
+
+void HistogramPool::create(const config::Parameters& p) {
+  auto name{p.get<std::string>("name")};
+  auto x_label{p.get<std::string>("xlabel")};
+  auto y_label{p.get<std::string>("ylabel")};
+  auto numeric_xbins{p.get<std::vector<double>>("xbins")};
+  auto category_xbins{p.get<std::vector<std::string>>("xcategories", {})};
+  auto numeric_ybins{p.get<std::vector<double>>("ybins", {})};
+  auto category_ybins{p.get<std::vector<std::string>>("ycategories", {})};
+
+  bool one_dim = (numeric_ybins.empty() and category_ybins.empty());
+  bool x_is_category = (not category_xbins.empty());
+  bool y_is_category = (not category_ybins.empty());
+  if (one_dim) {
+    // assume 1D histogram
+    if (x_is_category) {
+      create(name, x_label, category_xbins);
+    } else {
+      create(name, x_label, numeric_xbins);
+    }
+  } else {
+    if (x_is_category and y_is_category) {
+      create(name, x_label, category_xbins, y_label, category_ybins);
+    } else if (x_is_category and not y_is_category) {
+      create(name, x_label, category_xbins, y_label, numeric_ybins);
+    } else if (not x_is_category and y_is_category) {
+      create(name, x_label, numeric_xbins, y_label, category_ybins);
+    } else /* not x_is_category and not y_is_category */ {
+      create(name, x_label, numeric_xbins, y_label, numeric_ybins);
+    }
+  }
+}
+
+void HistogramPool::create(const std::string& name, const std::string& x_label,
+                           const std::vector<std::string>& categories,
+                           bool weighted) {
+  insert(
+      name,
+      [&]() {
```

Comment:
can you add a comment somewhere about why we are doing this lambda approach here, and then why we pass nothing in the body of it?